File: python-learning-path/practise/linear_search.py
"""
Exercise 16: Linear Search

ALGORITHM:
Sequentially check each element of the list until a match is found or the list ends.
Time Complexity: O(n)
"""

def linear_search(arr, target):
    # Iterate through indices of the list
    for index in range(len(arr)):
        # If element at current index matches target, return index
        if arr[index] == target:
            return index
            
    # If loop completes without finding target, return -1
    return -1

def main():
    data = [2, 3, 4, 10, 40]
    target = 10
    
    idx = linear_search(data, target)
    
    if idx != -1:
        print(f"🎯 Value {target} found at index {idx}")
    else:
        print(f"❌ Value {target} not found in the list.")

if __name__ == "__main__":
    main()

File: java-learning-path/practise/LinearSearch.java
/**
 * Exercise 16: Linear Search
 * 
 * ALGORITHM:
 * Simply scan the array from left to right.
 * Best for small or unsorted data.
 * Time Complexity: O(n)
 */
public class LinearSearch {
    public static void main(String[] args) {
        int[] arr = {2, 3, 4, 10, 40};
        int target = 10;
        
        int result = search(arr, target);
        
        if (result == -1)
            System.out.println("Element not found in searching.");
        else
            System.out.println("Element " + target + " found at index: " + result);
    }

    /**
     * @return index of target, or -1 if not found
     */
    public static int search(int[] arr, int target) {
        for (int i = 0; i < arr.length; i++) {
            // Found the value
            if (arr[i] == target) {
                return i;
            }
        }
        return -1; // Exhausted array without finding it
    }
}
